LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer
+{
+    using LoRaWan.NetworkServer.BasicsStation;
+    using Microsoft.Extensions.Caching.Memory;
+    using Microsoft.Extensions.Logging;
+    using System;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    public sealed class ConcentratorDeduplication : IConcentratorDeduplication, IDisposable
+    {
+        internal readonly MemoryCache Cache;
+
+        private readonly ILogger<IConcentratorDeduplication> Logger;
+        private readonly TimeSpan CacheEntryExpiration;
+
+        public ConcentratorDeduplication(
+            ILogger<IConcentratorDeduplication> logger,
+            int cacheEntryExpirationInMilliSeconds = 60_000)
+        {
+            this.Cache = new MemoryCache(new MemoryCacheOptions());
+            this.Logger = logger;
+            this.CacheEntryExpiration = TimeSpan.FromMilliseconds(cacheEntryExpirationInMilliSeconds);
+        }
+
+        public bool ShouldDrop(UpstreamDataFrame updf, StationEui stationEui)
+        {
+            if (updf == null) throw new ArgumentNullException(nameof(updf));
+
+            var key = CreateCacheKey(updf);
+            StationEui previousStation;
+
+            lock (this.Cache)
+            {
+                if (!this.Cache.TryGetValue(key, out previousStation))
+                {
+                    AddToCache(key, stationEui);
+                    return false;
+                }
+            }
+
+            if (previousStation == stationEui)
+            {
+                this.Logger.LogDebug($"Message received from the same DevAddr: {updf.DevAddr} as before, considered a resubmit.");
+                return false;
+            }
+
+            this.Logger.LogInformation($"Duplicate message detected from DevAddr: {updf.DevAddr}, dropping.");
+            return true;
+        }
+
+        internal static string CreateCacheKey(UpstreamDataFrame updf)
+        {
+            using var sha256 = SHA256.Create();
+            var key = sha256.ComputeHash(Encoding.UTF8.GetBytes(string.Join("", updf.DevAddr, updf.FrameCounter, updf.FRMPayload, updf.Mic)));
```

Comment:
The `string.Join` here will cause lots of allocations, including conversions that could be sensitive to culture. Can we not take advantage of the `Write` method on each primitive to write out the bytes?

Reply:
I worked on 2 variants: 
- [One is using the `Write` methods](0bc624bb9b8170bd143b63fdc20630b93f3ac4d8) when we have a primitive (e.g. the payload is a simple `string`). This implementation avoids new allocations by using `AsSpan()` (if my understanding is right) and also uses minimum amount of memory by allocating a buffer with the required length (again if my pointer arithmetics are correct). 
- [The second implementation uses a stringbuilder](bad5c28b9c7a5ebc8a56cd149af6d4f218a821e9) which allocates more memory (due to resizing) but is a lot simpler to understand and less error-prone. 

Imo we should go with the second implementation unless you think this is a critical path we must optimize heavily. Wdyt?

Reply:
> * [The second implementation uses a stringbuilder](bad5c28b9c7a5ebc8a56cd149af6d4f218a821e9) which allocates more memory (due to resizing) but is a lot simpler to understand and less error-prone.

The link was failing for me, but I think you meant to refer to commit bad5c28b9c7a5ebc8a56cd149af6d4f218a821e9?

Reply:
As discussed offline will go with the first option (modified).